grails-app/services/com/unifina/service/ProductService.groovy

```diff
@@ -14,8 +15,82 @@ class ProductService {
 	ApiService apiService
 	PermissionService permissionService
 	SubscriptionService subscriptionService
+	CassandraService cassandraService
 	Random random = ThreadLocalRandom.current()
 
+	static class StreamWithMessages {
```

Comment:
Why do we need to collect a list of messages? Looks like (line 70) that the list will always contain exactly 1 message anyway. Maybe just change the class to `StreamWithLatestMessage`:

```
static class StreamWithLatestMessage {
    Stream stream
    StreamrMessage latestMessage

    String toString() {
        // ...
    }
}
```
Also see comments below